File: levelbuilder.lua
local enum = require "util.enum"
local json = require "libs.json"
local Array = require "util.array"
local Levels = require "levels"
local Button = require "gui.button"

local LevelBuilder = {}

local Modes = enum {
    "TerrainBuilding",
    "RocketStartingLocation",
    "RocketLandingLocation",
    "Inspection"
}

local locationNotSetErrorMessage = [[
Rocket starting location or landing location were not set.
These are mandatory for the level to be saved.
]]

local buttonSchemes = {
    {
        name = "Inspection",
        mode = Modes.Inspection
    },
    {
        name = "Starting location",
        mode = Modes.RocketStartingLocation
    },
    {
        name = "Landing location",
        mode = Modes.RocketLandingLocation
    },
    {
        name = "Terrain",
        mode = Modes.TerrainBuilding
    }
}

function LevelBuilder:new(name)
    self.level = {
        name = name,
        terrainPoints = Array(),
        rocketStartingLocation = {
            x = nil, 
            y = nil,
            w = 200,
            h = 20
        },
        rocketLandingLocation = {
            x = nil, 
            y = nil,
            w = 200,
            h = 20
        }
    }

    do
        local startX = 10
        local buttonHeight = 30
        local buttonWidth = 200
        local buttonPadding = 20
        local distance = buttonWidth + buttonPadding

        self.mode = Modes.Inspection
        self.buttons = Array()
        self.newMode = false
        for i = 1, #buttonSchemes do
            local scheme = buttonSchemes[i]
            local x = startX + (i - 1) * distance
            local y = 0
            local w = buttonWidth
            local h = buttonHeight
            local text = scheme.name
            local callback = function()
                if self.mode ~= scheme.mode then
                    self.mode = scheme.mode
                    self.newMode= true
                end
            end
            self.buttons:push(Button(x, y, w, h, text, callback))
        end
    end
end

function LevelBuilder:update(dt)
    for button in self.buttons:iter() do
        button:update(dt)
    end
end

function LevelBuilder:mousepressed(mouseX, mouseY, button, istouch, presses)
    for button in self.buttons:iter() do
        button:mousepressed(mouseX, mouseY, button, istouch, presses)
    end
    if self.newMode == false then
        if self.mode == Modes.TerrainBuilding and button == 1 then
            self.level.terrainPoints:push(mouseX, mouseY)
        end
        if self.mode == Modes.RocketStartingLocation and button == 1 then
            self.level.rocketStartingLocation.x = mouseX
            self.level.rocketStartingLocation.y = mouseY
            self.mode = Modes.Inspection
        end
        if self.mode == Modes.RocketLandingLocation and button == 1 then
            self.level.rocketLandingLocation.x = mouseX
            self.level.rocketLandingLocation.y = mouseY
            self.mode = Modes.Inspection
        end
    end
    self.newMode = false
end

function LevelBuilder:keypressed(key, code, isrepeat)
    if key == "1" then
        self.mode = Modes.TerrainBuilding
    end
    if key == "2" then
        self.mode = Modes.RocketStartingLocation
    end
    if key == "3" then
        self.mode = Modes.RocketLandingLocation
    end
    if key == "4" then
        self.mode = Modes.Inspection
    end
    if key == "s" then
        if self.level.rocketStartingLocation.x and self.level.rocketStartingLocation.y and self.level.rocketLandingLocation.x and self.level.rocketLandingLocation.y then
            Levels.addLevel(self.level)
            Application.popState()
        else
            love.window.showMessageBox("Unable to save", locationNotSetErrorMessage, "error")
        end
    end
    if key == "escape" then
        love.event.quit()
    end
end

function LevelBuilder:draw()
    local mouseX, mouseY = love.mouse.getPosition()

    if self.level.terrainPoints:size() >= 2 then
        if self.mode == Modes.TerrainBuilding then
            self.level.terrainPoints:push(mouseX, mouseY)
        end

        love.graphics.line(self.level.terrainPoints:items())

        if self.mode == Modes.TerrainBuilding then
            self.level.terrainPoints:pop(2)
        end
    end

    local rect = self.level.rocketStartingLocation
    if self.mode == Modes.RocketStartingLocation then
        love.graphics.rectangle("fill", mouseX, mouseY, rect.w, rect.h)
    else
        if rect.x and rect.y then
            love.graphics.rectangle("fill", rect.x, rect.y, rect.w, rect.h)
        end
    end

    local rect = self.level.rocketLandingLocation
    if self.mode == Modes.RocketLandingLocation then
        love.graphics.rectangle("fill", mouseX, mouseY, rect.w, rect.h)
    else
        if rect.x and rect.y then
            love.graphics.rectangle("fill", rect.x, rect.y, rect.w, rect.h)
        end
    end

    for button in self.buttons:iter() do
        button:draw()
    end
end

local LevelBuilder__mt = {
    __index = LevelBuilder
}

return setmetatable(LevelBuilder, {
    __call = function(_, ...)
        local o = setmetatable({}, LevelBuilder__mt)
        o:new(...)
        return o
    end
})

File: src/levelbuilder.ts
import { KeyConstant } from "love.keyboard";

import { CampaignLevels } from "./campaignLevels";
import { GridRenderer } from "./gridrenderer";
import { Button } from "./gui";
import { LevelBuilderCamera } from "./levelbuildercamera";
import { Levels } from "./levels";

export interface Level {
    dataVersion: string;
    name: string;
    terrainPoints: number[][];
    rocketStartingLocation: Rectangle | undefined;
    rocketLandingLocation: Rectangle | undefined;
}

export interface CampaignLevelInfo {
    index: number;
}

enum Mode {
    TerrainBuilding,
    RocketStartingLocation,
    RocketLandingLocation,
    Inspection,
}

const locationNotSetErrorMessage = `
Rocket starting location or landing location were not set
These are mandatory for the level to be saved.
`;

const buttonSchemes = [
    {
        name: "Inspection",
        mode: Mode.Inspection,
    },
    {
        name: "Starting location",
        mode: Mode.RocketStartingLocation,
    },
    {
        name: "Landing location",
        mode: Mode.RocketLandingLocation,
    },
    {
        name: "Add Terrain",
        mode: Mode.TerrainBuilding,
    },
];

const rocketLocationWidth = 200;
const rocketLocationHeight = 20;

export class LevelBuilder implements GameState {
    private level: Level;
    private buttons: Button[];
    private mode: Mode;
    private newMode: boolean;
    private camera: LevelBuilderCamera;
    private gridRenderer: GridRenderer;
    private campignLevelInfo: CampaignLevelInfo | undefined;

    constructor(nameOrLevel: string | Level, campaignLevelInfo?: CampaignLevelInfo) {
        this.campignLevelInfo = campaignLevelInfo;

        if (type(nameOrLevel) == "string") {
            this.level = {
                dataVersion: "0.0.2",
                name: nameOrLevel as string,
                terrainPoints: [],
                rocketLandingLocation: undefined,
                rocketStartingLocation: undefined,
            };
        } else {
            const level = nameOrLevel as Level;
            this.level = level;
        }
        this.mode = Mode.Inspection;
        this.newMode = false;
        this.buttons = [];
        {
            const startX = 10;
            const buttonHeight = 30;
            const buttonWidth = 200;
            const buttonPadding = 20;
            const distance = buttonWidth + buttonPadding;
            for (let i = 0; i < buttonSchemes.length; i++) {
                let scheme = buttonSchemes[i];
                let x = startX + i * distance;
                let y = 0;
                let w = buttonWidth;
                let h = buttonHeight;
                let text = scheme.name;
                let callback = () => {
                    if (this.mode != scheme.mode) {
                        this.mode = scheme.mode;
                        this.newMode = true;

                        if (this.mode == Mode.TerrainBuilding) {
                            this.level.terrainPoints.push([]);
                        }
                    }
                };
                this.buttons.push(new Button(x, y, w, h, text, callback));
            }
        }
        {
            const w = 150;
            const h = 30;
            const x = love.graphics.getWidth() - w - 20;
            const y = 0;
            const text = "save";
            const callback = () => {
                if (!this.level.rocketLandingLocation || !this.level.rocketStartingLocation) {
                    love.window.showMessageBox("unable to save", locationNotSetErrorMessage, "error");
                } else {
                    if (campaignLevelInfo) {
                        CampaignLevels.addLevel(this.level);
                    } else {
                        Levels.addLevel(this.level);
                    }
                    Application.popState();
                }
            };
            this.buttons.push(new Button(x, y, w, h, text, callback));
        }
        this.camera = new LevelBuilderCamera();
        this.gridRenderer = new GridRenderer(250, this.camera);
    }

    getObjects() {
        return [this.camera];
    }

    update(dt: number) {
        for (const button of this.buttons) {
            button.update(dt);
        }
    }

    mousepressed(mouseX: number, mouseY: number, button: number, istouch: boolean, presses: number) {
        for (const b of this.buttons) {
            b.mousepressed(mouseX, mouseY, button, istouch, presses);
        }
        const [worldX, worldY] = this.camera.convertScreencoordinatesToWorldCoordinates(mouseX, mouseY);
        if (this.newMode == false) {
            switch (this.mode) {
                case Mode.TerrainBuilding:
                    if (button == 1) {
                        this.level.terrainPoints[this.level.terrainPoints.length - 1].push(worldX, worldY);
                    }
                    break;
                case Mode.RocketStartingLocation:
                    if (button == 1) {
                        if (this.level.rocketStartingLocation) {
                            this.level.rocketStartingLocation.x = worldX;
                            this.level.rocketStartingLocation.y = worldY;
                        } else {
                            this.level.rocketStartingLocation = {
                                x: worldX,
                                y: worldY,
                                w: rocketLocationWidth,
                                h: rocketLocationHeight,
                            };
                        }
                        this.mode = Mode.Inspection;
                    }
                    break;
                case Mode.RocketLandingLocation:
                    if (button == 1) {
                        if (this.level.rocketLandingLocation) {
                            this.level.rocketLandingLocation.x = worldX;
                            this.level.rocketLandingLocation.y = worldY;
                        } else {
                            this.level.rocketLandingLocation = {
                                x: worldX,
                                y: worldY,
                                w: rocketLocationWidth,
                                h: rocketLocationHeight,
                            };
                        }
                        this.mode = Mode.Inspection;
                    }
                    break;
            }
        }
        this.newMode = false;
    }

    keypressed(key: KeyConstant) {
        if (key == "escape") {
            if (this.mode == Mode.TerrainBuilding) {
                this.mode = Mode.Inspection;
            }
        }
    }

    draw() {
        this.camera.apply();
        this.gridRenderer.draw();

        const [worldX, worldY] = this.camera.convertScreencoordinatesToWorldCoordinates(...love.mouse.getPosition());
        for (let i = 0; i < this.level.terrainPoints.length; i++) {
            const points = this.level.terrainPoints[i];
            const isLast = i == this.level.terrainPoints.length - 1;
            if (points.length >= 2) {
                const terrainPreview = this.mode == Mode.TerrainBuilding && isLast;

                if (terrainPreview) {
                    points.push(worldX, worldY);
                }

                love.graphics.setColor(1, 0, 0, 1);
                love.graphics.setLineWidth(5);
                love.graphics.line(points as any);

                if (terrainPreview) {
                    points.pop();
                    points.pop();
                }
            }
        }

        love.graphics.setColor(1, 1, 1, 1);
        let rect = this.level.rocketStartingLocation;
        if (this.mode == Mode.RocketStartingLocation) {
            love.graphics.rectangle("fill", worldX, worldY, rocketLocationWidth, rocketLocationHeight);
        } else {
            if (rect) {
                love.graphics.rectangle("fill", rect.x, rect.y, rect.w, rect.h);
            }
        }

        rect = this.level.rocketLandingLocation;
        if (this.mode == Mode.RocketLandingLocation) {
            love.graphics.rectangle("fill", worldX, worldY, rocketLocationWidth, rocketLocationHeight);
        } else {
            if (rect) {
                love.graphics.rectangle("fill", rect.x, rect.y, rect.w, rect.h);
            }
        }

        love.graphics.origin();
        for (const button of this.buttons) {
            button.draw();
        }
        if (this.campignLevelInfo) {
            love.graphics.print(
                "campaign level " + tostring(this.campignLevelInfo.index),
                0,
                love.graphics.getHeight() - 15
            );
        }
    }
}
